## Data Cleaning Notebook

Main goal of this notebook is dimensionality reduction throughing dropping features with high rates of null values or ones that I deem to be irrelevant.

During this task, the codebook is being consistently referenced to help understand meaning of features.

Importing pandas, the only library needed for this task

In [1]:
import pandas as pd

Display files in local directory

In [2]:
ls

Codebook.pdf*                           globalterrorismdb_0617dist.xlsx*
Data Cleaning Notebook.ipynb            gtd1993_0617dist.xlsx*
Feature Engineering_EDA_Modeling.ipynb  gtd_13to16_0617dist.xlsx*
GTDDistributionletter2017.pdf*          gtd_70to94_0617dist.xlsx*
TermsofUse.pdf*                         gtd_95to12_0617dist.xlsx*
clean_data.csv


Read in excel file of data

In [3]:
data = pd.read_excel("globalterrorismdb_0617dist.xlsx")

View shape

In [4]:
data.shape

(170350, 135)

View names of features

In [5]:
data.columns

Index(['eventid', 'iyear', 'imonth', 'iday', 'approxdate', 'extended',
       'resolution', 'country', 'country_txt', 'region',
       ...
       'addnotes', 'scite1', 'scite2', 'scite3', 'dbsource', 'INT_LOG',
       'INT_IDEO', 'INT_MISC', 'INT_ANY', 'related'],
      dtype='object', length=135)

View memory usage of `data` variable

In [6]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 170350 entries, 0 to 170349
Columns: 135 entries, eventid to related
dtypes: datetime64[ns](1), float64(53), int64(24), object(57)
memory usage: 175.5+ MB


Display number of null values in each column

In [7]:
data.isnull().sum()

eventid                    0
iyear                      0
imonth                     0
iday                       0
approxdate            162886
extended                   0
resolution            168131
country                    0
country_txt                0
region                     0
region_txt                 0
provstate              14502
city                     446
latitude                4606
longitude               4606
specificity                4
vicinity                   0
location              122376
summary                66138
crit1                      0
crit2                      0
crit3                      0
doubtterr                  0
alternative           143719
alternative_txt       143719
multiple                   0
success                    0
suicide                    0
attacktype1                0
attacktype1_txt            0
                       ...  
propextent            109805
propextent_txt        109805
propvalue             134863
propcomment   

Create a list of column names that have 0 null values

In [8]:
non_null_cols = []

for col in data.columns:
    if data[col].isnull().sum() == 0:
        non_null_cols.append(col)

non_null_cols

['eventid',
 'iyear',
 'imonth',
 'iday',
 'extended',
 'country',
 'country_txt',
 'region',
 'region_txt',
 'vicinity',
 'crit1',
 'crit2',
 'crit3',
 'doubtterr',
 'multiple',
 'success',
 'suicide',
 'attacktype1',
 'attacktype1_txt',
 'targtype1',
 'targtype1_txt',
 'gname',
 'individual',
 'weaptype1',
 'weaptype1_txt',
 'property',
 'dbsource',
 'INT_LOG',
 'INT_IDEO',
 'INT_MISC',
 'INT_ANY']

In [10]:
len(non_null_cols)

31

There are 31 out of 175 columns with 0 values. 

Create a list of column names that have >0 null values

In [9]:
cols_with_nulls = []

for col in data.columns:
    if col not in non_null_cols:
        cols_with_nulls.append(col)

Pass in `cols_withs_nulls` into data and create a new dataframe

In [10]:
data_nulls = data[cols_with_nulls].copy()

In [11]:
data_nulls.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 170350 entries, 0 to 170349
Columns: 104 entries, approxdate to related
dtypes: datetime64[ns](1), float64(53), object(50)
memory usage: 135.2+ MB


In [12]:
data_nulls.columns

Index(['approxdate', 'resolution', 'provstate', 'city', 'latitude',
       'longitude', 'specificity', 'location', 'summary', 'alternative',
       ...
       'ransompaidus', 'ransomnote', 'hostkidoutcome', 'hostkidoutcome_txt',
       'nreleased', 'addnotes', 'scite1', 'scite2', 'scite3', 'related'],
      dtype='object', length=104)

Display total number of null values of each column

In [15]:
data_nulls.isnull().sum().sort_values(ascending = False)

gsubname3           170337
weapsubtype4_txt    170279
weapsubtype4        170279
weaptype4           170276
weaptype4_txt       170276
claimmode3_txt      170230
claimmode3          170230
gsubname2           170201
claim3              170087
guncertain3         170085
gname3              170081
divert              170037
attacktype3_txt     169976
attacktype3         169976
ransomnote          169879
ransompaidus        169863
ransomamtus         169854
claimmode2          169818
claimmode2_txt      169818
ransompaid          169643
corp3               169466
targsubtype3        169389
targsubtype3_txt    169389
natlty3             169344
natlty3_txt         169344
target3             169317
targtype3_txt       169316
targtype3           169316
ransomamt           169071
weapsubtype3        168843
                     ...  
propextent          109805
scite2              100765
ransom               94037
nperps               71124
nperpcap             69499
nwoundte             68749
n

Here I made the decision to not keep any column with more than 100k null values

In [17]:
cols_w_nulls_keep = data_nulls.isnull().sum().sort_values(ascending = False).iloc[-28:].index

Pass list of column names with <100k null values into `data_nulls`

In [18]:
data_nulls = data_nulls[cols_w_nulls_keep].copy()

In [19]:
data_nulls.head(1)

ransom  nperps  nperpcap  nwoundte  nkillter scite1 summary  claimed  \
0     0.0     NaN       NaN       NaN       NaN    NaN     NaN      NaN   

   nwoundus  nkillus     ...      targsubtype1_txt  longitude   latitude  \
0       NaN      NaN     ...        Named Civilian -69.951164  18.456792   

          natlty1_txt  natlty1       target1           city  guncertain1  \
0  Dominican Republic     58.0  Julio Guzman  Santo Domingo          0.0   

  ishostkid  specificity  
0       0.0          1.0  

[1 rows x 28 columns]

**This is the point where I go through inspect the values of each to inform my decision on whether or not to keep them**

View uniques of `ransom` variable

In [20]:
data_nulls.ransom.unique()

array([ 0.,  1., nan, -9.])

Fill in nulls of ransom with -9

In [21]:


data_nulls.ransom.fillna(-9., inplace=True)

Display counts of unique values in `nperps` column

In [22]:
data_nulls.nperps.value_counts(dropna=False)

-99.0       73051
NaN         71124
 1.0         7517
 2.0         5611
 3.0         2626
 4.0         2084
 5.0         1013
 6.0          737
 10.0         612
 20.0         609
 30.0         566
 50.0         528
 100.0        428
 15.0         409
 7.0          359
 8.0          347
 200.0        285
 40.0         275
 12.0         261
 25.0         178
 60.0         150
 9.0          139
 0.0          139
 300.0        134
 150.0        127
 70.0          86
 80.0          77
 24.0          68
 13.0          63
 11.0          54
            ...  
 28.0           2
 41.0           2
 58.0           2
 84.0           1
 25000.0        1
 3200.0         1
 750.0          1
 44.0           1
 64.0           1
 72.0           1
 48.0           1
 138.0          1
 230.0          1
 900.0          1
 370.0          1
 156.0          1
 190.0          1
 2500.0         1
 52.0           1
 68.0           1
 42.0           1
 49.0           1
 204.0          1
 53.0           1
 290.0    

Impute null values of nperps with -99.0

In [34]:
data_nulls.nperps.fillna(-99.0, inplace=True)

View unique values of `nperpcap` and impute null values with -99.0

In [35]:
data_nulls.nperpcap.unique()

array([-99. ,   1. ,   2. ,   0. ,   4. ,   3. ,   6. ,   5. ,   7. ,
         8. ,   9. ,  10. ,  26. ,  24. ,  13. ,  14. ,  45. ,  12. ,
        11. ,   2.5,  25. ,  32. ,  27. ,  -9. ,  15. ,  43. ,  21. ,
        50. , 406. ,  17. , 100. ,  60. ,  87. ,  55. ,  20. ,  29. ,
        76. ,  40. , 271. ,  28. ,  16. ,  19. ,  30. , 200. ,  41. ,
        18. , 126. ,  23. ,  22. ,  63. ])

In [24]:
data_nulls.nperpcap.fillna(-99., inplace=True)

View unique values of `nwoundte` and drop it

In [25]:
data_nulls.nwoundte.unique()

array([ nan,   0.,   1.,   4.,   2.,   3.,   5.,   6.,  50.,  10.,  16.,
        20.,   9.,  13.,  11., 200.,   7.,  15.,  31.,   8.,  12.,  25.,
        30.,  38.,  14.,  21.,  24.,  18.,  62.,  17., 150.,  22.,  23.,
        33.,  19.,  40.,  51., 101.,  60.,  35.,  53.,  26., 100.])

In [37]:
data_nulls.drop("nwoundte", axis = 1, inplace=True)

Drop `nkillter`

In [39]:
data_nulls.drop("nkillter", axis = 1, inplace=True)

In [27]:
data_nulls.head(1)

ransom  nperps  nperpcap  nwoundte scite1 summary  claimed  nwoundus  \
0     0.0     NaN     -99.0       NaN    NaN     NaN      NaN       NaN   

   nkillus weapdetail     ...      targsubtype1_txt  longitude   latitude  \
0      NaN        NaN     ...        Named Civilian -69.951164  18.456792   

          natlty1_txt natlty1       target1           city guncertain1  \
0  Dominican Republic    58.0  Julio Guzman  Santo Domingo         0.0   

   ishostkid  specificity  
0        0.0          1.0  

[1 rows x 27 columns]

Drop `scite1`

In [28]:
data_nulls.drop("scite1", axis = 1, inplace=True)

Display number of nulls in `summary` and then drop it

In [29]:
data_nulls.summary.isnull().sum()

66138

In [30]:
data_nulls.drop("summary", axis = 1, inplace=True)

View unique values in `claimed` and impute null values with -9.

In [31]:
data_nulls.claimed.unique()

array([nan,  0.,  1., -9.])

In [32]:
data_nulls.claimed.fillna(-9., inplace=True)

Display number of nulls in each column

In [40]:
data_nulls.isnull().sum()

ransom                  0
nperps                  0
nperpcap                0
claimed                 0
nwoundus            64710
nkillus             64459
weapdetail          59385
corp1               42566
weapsubtype1        19426
weapsubtype1_txt    19426
nwound              15325
provstate           14502
nkill                9682
targsubtype1         9345
targsubtype1_txt     9345
longitude            4606
latitude             4606
natlty1_txt          1394
natlty1              1394
target1               636
city                  446
guncertain1           379
ishostkid             178
specificity             4
dtype: int64

View data shape

In [41]:
data_nulls.shape

(170350, 24)

Drop the `nwoundus` and `nkillus` columns

In [42]:
data_nulls.drop(["nwoundus", "nkillus"], axis = 1, inplace=True)

Impute null values in `weapdetail` column with "no description"

In [43]:
data_nulls.weapdetail.fillna("no description", inplace=True)

Impute null values in `corp1` and `weapsubtype` with "unknown"

In [44]:
data_nulls.corp1.fillna("unknown", inplace=True)

In [45]:
data_nulls.weapsubtype1.fillna("unknown", inplace=True)

Drop `weapsubtype1_txt`

In [46]:
data_nulls.drop(["weapsubtype1_txt"], axis = 1, inplace=True)

Create dataframe `df` that consists of features with 0 nulls

In [47]:
df = data[non_null_cols].copy()

Concatenate along the columns the `df` and `data_nulls` dataframes

In [48]:
df = pd.concat([df, data_nulls], axis = 1)

View number of nulls in each column

In [49]:
df.isnull().sum()

eventid                 0
iyear                   0
imonth                  0
iday                    0
extended                0
country                 0
country_txt             0
region                  0
region_txt              0
vicinity                0
crit1                   0
crit2                   0
crit3                   0
doubtterr               0
multiple                0
success                 0
suicide                 0
attacktype1             0
attacktype1_txt         0
targtype1               0
targtype1_txt           0
gname                   0
individual              0
weaptype1               0
weaptype1_txt           0
property                0
dbsource                0
INT_LOG                 0
INT_IDEO                0
INT_MISC                0
INT_ANY                 0
ransom                  0
nperps                  0
nperpcap                0
claimed                 0
weapdetail              0
corp1                   0
weapsubtype1            0
nwound      

Drop any row that has a single null value

In [50]:
df.dropna(inplace=True)

Excess columns and nulls have been removed. Outputting clean data as a csv file.

In [51]:
df.to_csv("clean_data.csv")